### PR TITLE
Relax smart contract restrictions

### PR DESCRIPTION
**Pull request by Bargsteen**
## Purpose

Relax the restrictions on parameter size, return value size, and number of logs in smart contracts.
Depends on https://github.com/Concordium/concordium-base/pull/249

Solves part of https://github.com/Concordium/concordium-node/issues/563.

Most of the changes are in base PR, but this is what will be changed (from protocol version 5 and onward):
- Parameter size limit: 1kb -> 65kb
- Return value size limit: 16kb -> no limit (apart from energy)
- Number of logs: 64 -> no limit (apart from energy)
- Cost of parameters:
   - Of size <= 1kb: base cost + 1NRG / 1 **kilobyte** (same as before)
   - Of size > 1kb: base cost + 1NRG / 1 **byte**

## Changes

- Pass in the limits when running the wasm interpreter.

TODO:
- [ ] Add tests
   - [x] Tests for V1 contracts
   - [ ] Tests for V0 contracts
- [x] Changelog

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 1]
```diff
@@ -743,6 +743,9 @@ handleInitContract wtc initAmount modref initName param =
             -- with. Note that the deposit is already deducted at this point.
             senderAmount <- getCurrentAccountAvailableAmount senderAccount
 
+            -- Find setting for whether the number of logs and the size of return value should be limited.
+            let limitLogsAndRvs= Wasm.limitLogsAndReturnValues $ protocolVersion @(MPV m)
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/Scheduler.hs:747]:
```suggestion
            let limitLogsAndRvs = Wasm.limitLogsAndReturnValues $ protocolVersion @(MPV m)
```

File: concordium-consensus/src/Concordium/Scheduler/WasmIntegration/V1.hs [hunk 2]
```diff
@@ -293,7 +299,7 @@ data InvokeMethod =
 getInvokeMethod :: Get InvokeMethod
 getInvokeMethod = getWord8 >>= \case
   0 -> Transfer <$> get <*> get
-  1 -> Call <$> get <*> get <*> get <*> get
+  1 -> Call <$> get <*> Wasm.getParameterUnchecked <*> get <*> get -- TODO: Ensure that wasm interpreter checks this.
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/Scheduler/WasmIntegration/V1.hs:302]: There's a todo here!

Comment (abizjak) [inline concordium-consensus/src/Concordium/Scheduler/WasmIntegration/V1.hs:302]: It does.

**Review by abizjak — COMMENTED** [review]
Looks good.

Just the test missing and a changelog entry.

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 3]
```diff
@@ -1204,7 +1207,10 @@ handleContractUpdateV1 originAddr istance checkAndGetSender transferAmount recei
       withToContractAmountV1 sender istance transferAmount $ do
         foreignModel <- getRuntimeReprV1 model
         artifact <- liftLocal $ getModuleArtifact (GSWasm.miModule moduleInterface)
-        go [] =<< runInterpreter (return . WasmV1.applyReceiveFun artifact cm receiveCtx receiveName useFallback parameter transferAmount foreignModel)
+        let maxParameterLen = Wasm.maxParameterLen $ protocolVersion @(MPV m)
+        -- Find setting for whether the number of logs and the size of return value should be limited.
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/Scheduler.hs:1211]: Extremely nitpicky sorry. But I'm not sure "Find setting .." is the best way to phrase this. 
It's not really a `setting' to me (I would at least read this as something that could be toggled on/off/manipulated. 

I would probably vote for simply having the function called `isLogsAndReturnValuesLimited` and then the function is pretty much self explanatory from the context where you're providing the PV.

File: concordium-consensus/src/Concordium/Scheduler.hs [hunk 4]
```diff
@@ -1298,7 +1304,10 @@ handleContractUpdateV0 originAddr istance checkAndGetSender transferAmount recei
 
   model <- getRuntimeReprV0 (iiState istance)
   artifact <- liftLocal $ getModuleArtifact (GSWasm.miModule iface)
-  result <- runInterpreter (return . WasmV0.applyReceiveFun artifact cm receiveCtx receiveName parameter transferAmount model)
+  let maxParameterLen = Wasm.maxParameterLen $ protocolVersion @(MPV m)
+  -- Find setting for whether the number of logs and the size of return value should be limited.
```
Comment (EmilLa1) [inline concordium-consensus/src/Concordium/Scheduler.hs:1308]: Same as above (also extremely nitpicky sorry)

**Review by EmilLa1 — COMMENTED** [review]
Great job, looks good. 
I have a few very minor extremely nitpicky comments you can throw out the window if you find them too annoying :)

File: concordium-consensus/src/Concordium/Scheduler.hs (hunk 3 above)
Comment (Bargsteen) [reply concordium-consensus/src/Concordium/Scheduler.hs:1270]: I changed the comment and called it a day ;)

**Review by abizjak — APPROVED** [review]

**Comment by abizjak** [comment]
I will merge this now since there are some other PRs I want to get to.

I am quite confident the V0 is implemented correctly as well, but I will try to add a test for this tomorrow or Sunday in a separate PR.

**Pull request merged**